servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProvider.java

```diff
@@ -186,6 +205,23 @@ private static boolean isCompressed(Buffer buffer) throws SerializationException
             }
         }
 
+        private static CodedInputStream newCodedInputStream(final ByteBuffer[] buffers, final int lengthOfData) {
+            // Because we allocated a new internal ByteBuffer that will never be mutated we may just wrap it and
+            // enable aliasing to avoid an extra copying inside parser for a deserialized message.
+            final CodedInputStream in = unsafeWrap(mergeByteBuffers(buffers, lengthOfData)).newCodedInput();
```

Comment:
Wondering why we can not use `ByteString.unsafeWrap()` and aliasing for `toDeserialize.nioBufferCount() == 1` case also? We are pretty sure that the buffer will not be mutated between creation of `CodedInputStream` and usage by the parser.

Reply:
If `toDeserialize.nioBufferCount() == 1`, the `Buffer` will not create a new copy when we do `toDeserialize.toNioBuffer(...)`. Instead, it will return a `ByteBuffer` wrapper that shares the same internal array.
Because users have a way to access the original `Buffer` using http filters, we can not guarantee that they do not modify the internal array. Therefore, by default we need to be safe. However, we can provide a configuration option for `ProtoBufSerializationProviderBuilder` and `GrpcClientBuilder/GrpcServerBuilder` to do zero-copying deserialization when users sure that they do not touch the payload body in any other way.